File: server_litestar_piccolo/app/models/items.py
"""PyDantic models."""

from __future__ import annotations

from datetime import date, datetime  # noqa: TC003
from typing import Annotated, Literal

from pydantic import (
    BaseModel,
    BeforeValidator,
    ConfigDict,
    Field,
)

from app.classes.classes import Conclusions, Decisions


class IdModel(BaseModel):
    """ItemModel schema."""

    id: int


class DateIdModel(IdModel):
    """DateIdModel schema."""

    created_at: datetime
    updated_at: datetime


class PrevIn(BaseModel):
    """Previous in schema."""

    model_config = ConfigDict(str_strip_whitespace=True)

    surname: str = Field(max_length=255)
    firstname: Annotated[str | None, Field(default=None, max_length=255)]
    patronymic: Annotated[str | None, Field(default=None, max_length=255)]
    changed: Annotated[
        str | None,
        Field(default=None, max_length=4),
        BeforeValidator(str),
    ]
    reason: str | None = None
    item: Literal["previous"]


class PrevOut(PrevIn, IdModel):
    """Previous out schema."""


class EducationIn(BaseModel):
    """Education in schema."""

    model_config = ConfigDict(str_strip_whitespace=True)

    view: Annotated[str | None, Field(default=None, max_length=255)]
    institution: Annotated[str, Field(max_length=255)]
    finished: Annotated[str | None, Field(default=None), BeforeValidator(str)]
    specialty: str | None = None
    item: Literal["educations"]


class EducationOut(EducationIn, IdModel):
    """Educations schema."""


class StaffIn(BaseModel):
    """Staffs schema."""

    model_config = ConfigDict(str_strip_whitespace=True)

    position: str
    department: str | None = None
    item: Literal["staffs"]


class StaffOut(StaffIn, IdModel):
    """Staffs out schema."""


class DocumentIn(BaseModel):
    """Document in schema."""

    model_config = ConfigDict(str_strip_whitespace=True)

    view: str | None = "Паспорт"
    series: str | None = None
    digits: Annotated[str, Field(max_length=12)]
    agency: Annotated[str | None, Field(default=None, max_length=255)]
    issue: date | None = None
    item: Literal["documents"]


class DocumentOut(DocumentIn, IdModel):
    """Document out schema."""


class AddressIn(BaseModel):
    """Address in schema."""

    model_config = ConfigDict(str_strip_whitespace=True, str_max_length=255)

    view: str
    address: str
    item: Literal["addresses"]


class AddressOut(AddressIn, IdModel):
    """Address out schema."""


class ContactIn(BaseModel):
    """Contacts in schema."""

    model_config = ConfigDict(str_strip_whitespace=True, str_max_length=255)

    view: str
    contact: str
    item: Literal["contacts"]


class ContactOut(ContactIn, IdModel):
    """Contacts out schema."""


class WorkplaceIn(BaseModel):
    """Workplaces in schema."""

    model_config = ConfigDict(str_strip_whitespace=True)

    now_work: Annotated[bool | None, BeforeValidator(bool)]
    starts: date | None = None
    finished: Annotated[
        date | None | str,
        BeforeValidator(lambda v: None if isinstance(v, str) else v),
    ]
    workplace: Annotated[str, Field(max_length=255)]
    address: Annotated[str | None, Field(None, max_length=255)]
    position: Annotated[str, Field(max_length=255)]
    reason: str | None = None
    item: Literal["workplaces"]


class WorkplaceOut(WorkplaceIn, IdModel):
    """Workplace out schema."""


class AffilationIn(BaseModel):
    """Affilation in schema."""

    model_config = ConfigDict(str_strip_whitespace=True)

    view: Annotated[str, Field(max_length=255)]
    organization: Annotated[str, Field(max_length=255)]
    inn: Annotated[str | None, Field(None, max_length=12)]
    item: Literal["affilations"]


class AffilationOut(AffilationIn, IdModel):
    """Affilations out schema."""


class CheckIn(DateIdModel):
    """Check in schema."""

    model_config = ConfigDict(use_enum_values=True)

    workplace: str | None = None
    document: str | None = None
    inn: str | None = None
    debt: str | None = None
    bankruptcy: str | None = None
    bki: str | None = None
    courts: str | None = None
    affilation: str | None = None
    terrorist: str | None = None
    mvd: str | None = None
    internet: str | None = None
    cronos: str | None = None
    cros: str | None = None
    addition: str | None = None
    comment: str | None = None
    conclusion: Conclusions
    person_id: int | None = None
    item: Literal["checks"]


class CheckOut(CheckIn, DateIdModel):
    """Checks out schema."""


class PoligrafIn(DateIdModel):
    """Poligraf in schema."""

    model_config = ConfigDict(use_enum_values=True)

    theme: str
    results: str
    conclusion: Decisions
    item: Literal["poligrafs"]


class PoligrafOut(PoligrafIn, DateIdModel):
    """Poligraf out schema."""


class InvestigationIn(BaseModel):
    """Investigations in schema."""

    theme: str
    info: str
    item: Literal["investigations"]


class InvestigationOut(InvestigationIn, DateIdModel):
    """Investigation out schema."""


class InquiryIn(BaseModel):
    """Inquiries in schema."""

    info: str
    initiator: str
    item: Literal["inquiries"]


class InquiryOut(InquiryIn, DateIdModel):
    """Inquiries out schema."""


ItemTypeIn = Annotated[
    AddressIn
    | AffilationIn
    | CheckIn
    | ContactIn
    | DocumentIn
    | EducationIn
    | InquiryIn
    | InvestigationIn
    | PrevIn
    | PoligrafIn
    | StaffIn
    | WorkplaceIn,
    Field(discriminator="item"),
]


class ItemModelIn(BaseModel):
    """Validation class."""

    item: ItemTypeIn


ItemTypeOut = Annotated[
    AddressOut
    | AffilationOut
    | CheckOut
    | ContactOut
    | DocumentOut
    | EducationOut
    | InquiryOut
    | InvestigationOut
    | PrevOut
    | PoligrafOut
    | StaffOut
    | WorkplaceOut,
    Field(discriminator="item"),
]


class ItemModelOut(BaseModel):
    """Validation class."""

    item: ItemTypeOut


class ItemsOutModels(BaseModel):
    """Validation class."""

    staffs: list[StaffOut]
    educations: list[EducationOut]
    workplaces: list[WorkplaceOut]
    documents: list[DocumentOut]
    addresses: list[AddressOut]
    contacts: list[ContactOut]
    affilations: list[AffilationOut]
    previous: list[PrevOut]
    checks: list[CheckOut]
    poligrafs: list[PoligrafOut]
    investigations: list[InvestigationOut]
    inquiries: list[InquiryOut]

File: web_nuxt/app/schema/items.ts
import { h } from "vue";
import { localStr } from "@/utils";
import { Conclusions, Decisions } from "@/types";
import type { Items, ItemFields } from "@/types";

export const person = [
  { key: "surname", label: "Фамилия" },
  { key: "firstname", label: "Имя" },
  { key: "patronymic", label: "Отчество" },
  {
    key: "birthday",
    label: "Дата рождения",
    div: (row) => localStr(row.birthday),
  },
  { key: "birthplace", label: "Место рождения" },
  { key: "citizenship", label: "Гражданство" },
  { key: "dual", label: "Двойное гражданство" },
  { key: "snils", label: "СНИЛС" },
  { key: "inn", label: "ИНН" },
  { key: "marital", label: "Семейное положение" },
  { key: "addition", label: "Дополнительная информация" },
  {
    key: "created",
    label: "Дата записи",
    div: (row) => localStr(row.updated_at),
  },
  { key: "destination", label: "Материалы проверок", slot: true },
] as ItemFields<Items["person"]>[];

const addresses = [
  { key: "view", label: "Вид адреса" },
  { key: "address", label: "Адрес" },
] as ItemFields<Items["addresses"]>[];

const affilations = [
  { key: "view", label: "Вид участия" },
  { key: "organization", label: "Организация" },
  { key: "inn", label: "ИНН" },
  { key: "activity", label: "Деятельность" },
] as ItemFields<Items["affilations"]>[];

const checks = [
  { key: "workplace", label: "Проверка по местам работы" },
  { key: "document", label: "Проверка документов" },
  { key: "debt", label: "Проверка задолженностей" },
  { key: "bankruptcy", label: "Проверка банкротства" },
  { key: "bki", label: "Проверка Кредитной истории" },
  { key: "courts", label: "Проверка судебных дел" },
  { key: "affilation", label: "Проверка аффилированности" },
  { key: "terrorist", label: "Проверка в списке террористов" },
  { key: "internet", label: "Проверка в открытых источниках" },
  { key: "cronos", label: "Проверка в Кронос" },
  { key: "addition", label: "Дополнительная информация" },
  { key: "comment", label: "Комментарий" },
  {
    key: "conclusion",
    label: "Результат",
    component: (row) => {
      const color =
        row.conclusion === Conclusions.agreed
          ? "bg-green-500"
          : row.conclusion === Conclusions.cancel
            ? "bg-gray-500"
            : row.conclusion === Conclusions.comments
              ? "bg-blue-500"
              : "bg-red-500";
      return h(
        "span",
        {
          class: [
            "flex inline-flex items-center py-1 px-2 rounded-md text-sm text-white",
            color,
          ],
        },
        row.conclusion,
      );
    },
  },
  {
    key: "created",
    label: "Дата записи",
    div: (row) => localStr(row.updated_at),
  },
] as ItemFields<Items["checks"]>[];

const contacts = [
  { key: "view", label: "Вид контакта" },
  { key: "contact", label: "Контакт" },
] as ItemFields<Items["contacts"]>[];

const documents = [
  { key: "view", label: "Вид документа" },
  { key: "series", label: "Серия документа" },
  { key: "digits", label: "Номер документа" },
  { key: "agency", label: "Кем выдан" },
  { key: "issue", label: "Дата выдачи", div: (row) => localStr(row.issue) },
] as ItemFields<Items["documents"]>[];

const educations = [
  { key: "view", label: "Вид образования" },
  { key: "institution", label: "Учебное заведение" },
  { key: "finished", label: "Год окончания" },
  { key: "specialty", label: "Специальность" },
] as ItemFields<Items["educations"]>[];

const inquiries = [
  { key: "info", label: "Информация" },
  { key: "initiator", label: "Инициатор" },
  {
    key: "created",
    label: "Дата записи",
    div: (row) => localStr(row.updated_at),
  },
] as ItemFields<Items["inquiries"]>[];

const investigations = [
  { key: "theme", label: "Тема проверки" },
  { key: "info", label: "Информация" },
  {
    key: "created",
    label: "Дата записи",
    div: (row) => localStr(row.updated_at),
  },
] as ItemFields<Items["investigations"]>[];

const poligrafs = [
  { key: "theme", label: "Тема проверки" },
  { key: "results", label: "Результат" },
  {
    key: "conclusion",
    label: "Результат",
    component: (row) => {
      const color =
        row.conclusion === Decisions.agreed
          ? "bg-green-500"
          : row.conclusion === Decisions.cancel
            ? "bg-gray-500"
            : row.conclusion === Decisions.comments
              ? "bg-blue-500"
              : "bg-red-500";
      return h(
        "span",
        {
          class: [
            "flex inline-flex items-center py-1 px-2 rounded-md text-sm text-white",
            color,
          ],
        },
        row.conclusion,
      );
    },
  },
  {
    key: "created",
    label: "Дата записи",
    div: (row) => localStr(row.updated_at),
  },
] as ItemFields<Items["poligrafs"]>[];

const previous = [
  { key: "surname", label: "Фамилия" },
  { key: "firstname", label: "Имя" },
  { key: "patronymic", label: "Отчество" },
  { key: "changed", label: "Год изменения" },
  { key: "reason", label: "Причина" },
] as ItemFields<Items["previous"]>[];

const staffs = [
  { key: "position", label: "Должность" },
  { key: "department", label: "Подразделение" },
] as ItemFields<Items["staffs"]>[];

const workplaces = [
  { key: "starts", label: "Начало работы", div: (row) => localStr(row.starts) },
  {
    key: "finished",
    label: "Окончание работы",
    div: (row) => localStr(row.finished),
  },
  { key: "workplace", label: "Место работы" },
  { key: "position", label: "Должность" },
  { key: "address", label: "Адрес организации" },
  { key: "reason", label: "Причина увольнения" },
] as ItemFields<Items["workplaces"]>[];

export const itemsFields = {
  addresses,
  affilations,
  checks,
  contacts,
  documents,
  educations,
  inquiries,
  investigations,
  person,
  poligrafs,
  previous,
  staffs,
  workplaces,
} as { [key in keyof Items]: ItemFields<Items[keyof Items]>[] };
